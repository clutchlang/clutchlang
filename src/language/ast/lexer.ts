export * from './lexer/token';
export * from './lexer/tokenizer';
