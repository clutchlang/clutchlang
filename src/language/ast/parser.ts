import * as tokens from './token';

/**
 * Represents an unresolved node in the abstract syntax tree.
 */
export abstract class AstNode {
  public abstract accept<R, C>(
    visitor: AstVisitor<R, C>,
    context?: C
  ): R | undefined;

  /**
   * First token that represents this AST node.
   */
  public abstract get first(): tokens.Token;

  /**
   * Last token that represents this AST node.
   */
  public abstract get last(): tokens.Token;
}

/**
 * Represents a simple AST node that originates from a single token.
 */
export abstract class SimpleNode extends AstNode {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an operator node parsed from a single token.
 */
export class Operator extends SimpleNode {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitOperator(this, context);
  }
}

/**
 * Represents a marker type of AST node for all statements.
 */
export abstract class Statement extends AstNode {}

/**
 * Represents a marker type of AST node for all expressions.
 */
export abstract class Expression extends Statement {}

/**
 * Represents a simple AST node that is also an expression.
 */
export abstract class SimpleExpression extends Expression {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an expression that utilizes an operator.
 */
export abstract class OperatorExpression extends Expression {
  constructor(
    public readonly operator: Operator,
    public readonly target: Expression
  ) {
    super();
  }
}

/**
 * Represents an expression in the form of `<OP><EXPR>`.
 */
export class PrefixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPrefixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.operator.first;
  }

  public get last(): tokens.Token {
    return this.target.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP>`.
 */
export class PostfixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPostfixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.operator.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP><EXPR>`.
 */
export class BinaryExpression extends OperatorExpression {
  constructor(
    operator: Operator,
    public readonly left: Expression,
    public readonly right: Expression
  ) {
    super(operator, left);
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitBinaryExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.left.first;
  }

  public get last(): tokens.Token {
    return this.right.last;
  }
}

/**
 * Represents an expression in the form of `<TARGET>.<PROPERTY>`.
 */
export class PropertyExpression extends Expression {
  constructor(
    public readonly target: Expression,
    public readonly property: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPropertyExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.property.last;
  }
}

/**
 * Represents a list of arguments passed to a function call.
 */
export class ArgumentList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly args: Expression[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitArgumentList(this, context);
  }
}

/**
 * Represents an expression for a function call.
 *
 * May _optionally_ contain a @member target.
 */
export class CallExpression extends Expression {
  constructor(
    public readonly name: Identifier,
    public readonly args: ArgumentList,
    public readonly target?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitCallExpression(this, context);
  }

  public get first(): tokens.Token {
    return (this.target || this.name).first;
  }

  public get last(): tokens.Token {
    return this.args.last;
  }
}

export class ConditionalExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly condition: Expression,
    public readonly body: Expression | StatementBlock,
    public readonly elseCondition?: Expression,
    public readonly elseBody?: Expression | StatementBlock
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitConditionalExpression(this, context);
  }

  public get last(): tokens.Token {
    return (this.elseBody || this.body).last;
  }
}

export class GroupExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression: Expression,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitGroupExpression(this, context);
  }
}

/**
 * Represents an expression for a literal identifier (name).
 */
export class Identifier extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitIdentifier(this, context);
  }

  public get name(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal boolean.
 */
export class LiteralBoolean extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralBoolean(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal number.
 */
export class LiteralNumber extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralNumber(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal string.
 */
export class LiteralString extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralString(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents a variable declared anywhere in scope.
 */
export class VariableDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly type?: Identifier,
    public readonly value?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitVariableDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return (this.value || this.type || this.name).last;
  }
}

/**
 * Represents a list of parameters for a function or closure definition.
 */
export class ParameterList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly params: VariableDeclaration[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitParameterList(this, context);
  }
}

/**
 * Represents a series of statements.
 */
export class StatementBlock extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly statements: Statement[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitStatementBlock(this, context);
  }
}

/**
 * Represents a return statement.
 */
export class ReturnStatement extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitReturnStatement(this, context);
  }

  public get last(): tokens.Token {
    return this.expression ? this.expression.last : this.first;
  }
}

/**
 * Represents a function declared anywhere in scope.
 */
export class FunctionDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly params: ParameterList,
    public readonly body: Expression | StatementBlock,
    public readonly returnType?: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitFunctionDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return this.body.last;
  }
}

/**
 * Represents a type declared anywhere in scope.
 */
export class TypeDeclaration extends AstNode {
  constructor(
    public readonly modifiers: tokens.Token[],
    public readonly name: Identifier,
    public readonly members: Array<FunctionDeclaration | VariableDeclaration>,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitTypeDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.modifiers.length ? this.modifiers[0] : this.name.first;
  }
}

/**
 *
 */
export class ModuleRoot extends AstNode {
  constructor(
    public readonly declarations: Array<
      FunctionDeclaration | TypeDeclaration | VariableDeclaration
    >
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitModuleRoot(this, context);
  }

  public get first(): tokens.Token {
    return this.declarations[0].first;
  }

  public get last(): tokens.Token {
    return this.declarations[this.declarations.length - 1].last;
  }
}

export class AstParser {}

/**
 * An AST visitor that by default does nothing visiting an AST node.
 *
 * It is intended to be a superclass for classes that use the visitor pattern
 * primarily as a dispatch mechanism (and do not need to recursively need to
 * visit the whole structure) and only need to visit a small number of node
 * types.
 */
export class AstVisitor<R, C> {
  public visitArgumentList(_: ArgumentList, __?: C): R | undefined {
    return;
  }

  public visitBinaryExpression(_: BinaryExpression, __?: C): R | undefined {
    return;
  }

  public visitCallExpression(_: CallExpression, __?: C): R | undefined {
    return;
  }

  public visitConditionalExpression(
    _: ConditionalExpression,
    __?: C
  ): R | undefined {
    return;
  }

  public visitFunctionDeclaration(
    _: FunctionDeclaration,
    __?: C
  ): R | undefined {
    return;
  }

  public visitGroupExpression(_: GroupExpression, __?: C): R | undefined {
    return;
  }

  public visitIdentifier(_: Identifier, __?: C): R | undefined {
    return;
  }

  public visitLiteralBoolean(_: LiteralBoolean, __?: C): R | undefined {
    return;
  }

  public visitLiteralNumber(_: LiteralNumber, __?: C): R | undefined {
    return;
  }

  public visitLiteralString(_: LiteralString, __?: C): R | undefined {
    return;
  }

  public visitModuleRoot(_: ModuleRoot, __?: C): R | undefined {
    return;
  }

  public visitOperator(_: Operator, __?: C): R | undefined {
    return;
  }

  public visitParameterList(_: ParameterList, __?: C): R | undefined {
    return;
  }

  public visitPrefixExpression(_: PrefixExpression, __?: C): R | undefined {
    return;
  }

  public visitPostfixExpression(_: PostfixExpression, __?: C): R | undefined {
    return;
  }

  public visitPropertyExpression(_: PropertyExpression, __?: C): R | undefined {
    return;
  }

  public visitReturnStatement(_: ReturnStatement, __?: C): R | undefined {
    return;
  }

  public visitStatementBlock(_: StatementBlock, __?: C): R | undefined {
    return;
  }

  public visitTypeDeclaration(_: TypeDeclaration, __?: C): R | undefined {
    return;
  }

  public visitVariableDeclaration(
    _: VariableDeclaration,
    __?: C
  ): R | undefined {
    return;
  }
}
