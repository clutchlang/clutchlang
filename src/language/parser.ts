export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
