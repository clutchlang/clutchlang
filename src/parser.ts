export { Codes } from './parser/codes';
export { Lexer } from './parser/lexer';
export { Token } from './parser/token';
export { Tokens } from './parser/tokens';
