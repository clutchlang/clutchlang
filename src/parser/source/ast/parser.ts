import {
  RegExpToken,
  StringToken,
  SymbolToken,
  Token,
  TokenKind,
} from '../tokenizer/tokens';
import {
  AstCompilationUnit,
  AstExpression,
  AstFunctionDeclaration,
  AstInvocationExpression,
  AstLiteralBoolean,
  AstLiteralIdentifier,
  AstLiteralNumber,
  AstLiteralString,
  AstParenthesizedExpression,
} from './node';
import { TokenScanner } from './scanner';

export class AstParser {
  constructor(private readonly scanner: TokenScanner) {}

  public parseCompilationUnit(): AstCompilationUnit {
    const results: AstFunctionDeclaration[] = [];
    while (!this.scanner.isDone) {
      // TODO: Loosen once alternatives are supported.
      this.scanRequired(RegExpToken.Identifier);
      results.push(this.parseFunctionDeclaration());
    }
    return new AstCompilationUnit(results);
  }

  /**
   * Returns a parsed expression, or undefined if there is none.
   */
  private parseExpression(): AstExpression | undefined {
    const token = this.scanner.peek();
    switch (token.kind) {
      case SymbolToken.LParen:
        this.scanRequired(SymbolToken.LParen);
        const expressions = this.parseExpressions();
        this.scanRequired(SymbolToken.RParen);
        return new AstParenthesizedExpression(expressions);
      case RegExpToken.Identifier:
        if (this.scanner.peek(1).kind === SymbolToken.LParen) {
          const identifier = this.scanner.read();
          this.scanRequired(SymbolToken.LParen);
          const invocation = new AstInvocationExpression(
            new AstLiteralIdentifier(identifier),
            this.parseExpressions()
          );
          this.scanRequired(SymbolToken.RParen);
          return invocation;
        }
        return new AstLiteralIdentifier(this.scanner.read());
      case RegExpToken.LiteralBoolean:
        return new AstLiteralBoolean(this.scanner.read());
      case RegExpToken.LiteralNumber:
        return new AstLiteralNumber(this.scanner.read());
      case RegExpToken.LiteralString:
        return new AstLiteralString(this.scanner.read());
      default:
        return undefined;
    }
  }

  private parseFunctionDeclaration(): AstFunctionDeclaration {
    const identifier = this.scanner.lastMatch[0];
    this.scanRequired(StringToken.Arrow);
    if (this.scanOptional(SymbolToken.LCurly)) {
      const fn = new AstFunctionDeclaration(
        identifier,
        this.parseExpressions()
      );
      this.scanRequired(SymbolToken.RCurly);
      return fn;
    }
    const expression = this.parseExpression()!;
    return new AstFunctionDeclaration(identifier, [expression]);
  }

  private parseExpressions(): AstExpression[] {
    const expressions: AstExpression[] = [];
    let expression = this.parseExpression();
    while (expression) {
      expressions.push(expression);
      expression = this.parseExpression();
    }
    return expressions;
  }

  private scanOptional(token: TokenKind): Token | undefined {
    if (this.scanner.scan(token)) {
      return this.scanner.lastMatch[0];
    }
    return undefined;
  }

  private scanRequired(token: TokenKind): void {
    /* istanbul ignore next */
    if (!this.scanner.scan(token)) {
      // The lexer should prevent this happening in a typical program.
      /* istanbul ignore next */
      const next = this.scanner.peek().kind;
      /* istanbul ignore next */
      throw new SyntaxError(`Expected ${token.name} got ${next.name}.`);
    }
  }
}
