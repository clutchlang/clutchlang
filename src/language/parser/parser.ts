import { IToken, TokenKind } from '../lexer';
import { AstNodeFactory } from './factory';
import { GroupExpression } from './nodes/expressions';
import { Expression } from './nodes/nodes';
import { Operator, Precedence } from './nodes/operators';

export class ClutchParser {
  /**
   * Position within @member tokens.
   */
  private position = 0;

  constructor(
    private readonly tokens: IToken[],
    private readonly factory = new AstNodeFactory()
  ) {}

  /**
   * Parses and returns an expression.
   *
   * Uses [recursive descent parsing][1].
   *
   * [1]: http://craftinginterpreters.com/parsing-expressions.html#recursive-descent-parsing
   */
  public parseExpression(precedence = Precedence.Assignment): Expression {
    let expr!: Expression;
    let operator!: IToken;

    // An inline helper for parsing binary expressions.
    const parseBinary = (kinds: TokenKind[], next?: Precedence): Expression => {
      expr = this.parseExpression(next);
      while (this.match(...kinds)) {
        operator = this.peek(-1);
        expr = this.factory.createBinaryExpression(
          expr,
          this.parseBinaryOperator(operator.lexeme),
          operator,
          this.parseExpression(next)
        );
      }
      return expr;
    };

    // An inline helper for parsing unary expressions.
    const parseUnary = (
      prefix: boolean,
      kinds: TokenKind[],
      next?: Precedence
    ): Expression => {
      if (!prefix) {
        // Special case postfix.
        operator = this.peek(1);
        if (kinds.some(e => e === operator.lexeme)) {
          return this.factory.createUnaryExpression(
            this.parseExpression(next),
            this.parseUnaryOperator(operator.lexeme, prefix),
            operator,
            prefix
          );
        }
      }
      if (this.match(...kinds)) {
        operator = this.peek(-1);
        return this.factory.createUnaryExpression(
          this.parseExpression(next),
          this.parseUnaryOperator(operator.lexeme, prefix),
          operator,
          prefix
        );
      }
      return this.parseExpression(next);
    };

    switch (precedence) {
      case Precedence.Assignment:
        return parseBinary(
          [
            TokenKind.ASSIGN,
            TokenKind.PLUS_BY,
            TokenKind.MINUS_BY,
            TokenKind.STAR_BY,
            TokenKind.SLASH_BY,
            TokenKind.MODULUS_BY,
          ],
          Precedence.Disjunction
        );
      case Precedence.Disjunction:
        return parseBinary([TokenKind.OR], Precedence.Conjunction);
      case Precedence.Conjunction:
        return parseBinary([TokenKind.AND], Precedence.Equality);
      case Precedence.Equality:
        return parseBinary(
          [
            TokenKind.EQUALS,
            TokenKind.NOT_EQUALS,
            TokenKind.IDENTICAL,
            TokenKind.NOT_IDENTICAL,
          ],
          Precedence.Comparison
        );
      case Precedence.Comparison:
        return parseBinary(
          [
            TokenKind.LESS_THAN,
            TokenKind.GREATER_THAN,
            TokenKind.LESS_THAN_OR_EQUAL,
            TokenKind.GREATER_THAN_OR_EQUAL,
          ],
          Precedence.Additive
        );
      case Precedence.Additive:
        return parseBinary(
          [TokenKind.PLUS, TokenKind.MINUS],
          Precedence.Multiplicative
        );
      case Precedence.Multiplicative:
        return parseBinary(
          [TokenKind.STAR, TokenKind.SLASH, TokenKind.MODULUS],
          Precedence.Prefix
        );
      case Precedence.Prefix:
        return parseUnary(
          true,
          [
            TokenKind.MINUS,
            TokenKind.PLUS,
            TokenKind.INCREMENT,
            TokenKind.DECREMENT,
            TokenKind.NEGATE,
          ],
          Precedence.Accessor
        );
      case Precedence.Accessor:
        return parseBinary([TokenKind.PERIOD], Precedence.Postfix);
      case Precedence.Postfix:
        return parseUnary(
          false,
          [TokenKind.INCREMENT, TokenKind.DECREMENT],
          Precedence.Literal
        );
      default:
        if (this.match(TokenKind.LEFT_PAREN)) {
          return this.parseGroupExpression();
        }
        return this.parseLiteral();
    }
  }

  private parseGroupExpression(): GroupExpression {
    const leftParen = this.peek(-1);
    const expression = this.parseExpression();
    const rightParen = this.advance();
    return new GroupExpression(leftParen, rightParen, expression);
  }

  private parseLiteral(): Expression {
    if (this.match(TokenKind.NUMBER)) {
      return this.factory.createLiteralNumber(this.peek(-1));
    }
    if (this.match(TokenKind.STRING)) {
      return this.factory.createLiteralString(this.peek(-1));
    }
    if (this.match(TokenKind.FALSE, TokenKind.TRUE)) {
      return this.factory.createLiteralBoolean(this.peek(-1));
    }
    if (!this.match(TokenKind.IDENTIFIER)) {
      // TODO: Test.
      /* istanbul ignore next */
      throw new SyntaxError(
        `Expected literal got "${this.peek().lexeme}" (${this.peek().kind}).`
      );
    }
    const token = this.peek(-1);
    return this.factory.createSimpleName(token);
  }

  private parseBinaryOperator(lexeme: string): Operator {
    switch (lexeme) {
      case '.':
        return Operator.Accessor;
      case '*':
        return Operator.Multiply;
      case '/':
        return Operator.Divide;
      case '%':
        return Operator.Modulus;
      case '+':
        return Operator.Add;
      case '-':
        return Operator.Subtract;
      case '<':
        return Operator.Less;
      case '>':
        return Operator.Greater;
      case '<=':
        return Operator.LessOrEqual;
      case '>=':
        return Operator.GreaterOrEqual;
      case '==':
        return Operator.Equal;
      case '!=':
        return Operator.NotEqual;
      case '===':
        return Operator.Identical;
      case '!==':
        return Operator.NotIdentical;
      case '&&':
        return Operator.And;
      case '||':
        return Operator.Or;
      case '=':
        return Operator.Assign;
      case '+=':
        return Operator.AddAssign;
      case '-=':
        return Operator.SubtractAssign;
      case '*=':
        return Operator.MultiplyAssign;
      case '/=':
        return Operator.DivideAssign;
      case '%=':
        return Operator.ModulusAssign;
      default:
        // TODO: Test.
        /* istanbul ignore next */
        throw new SyntaxError(`Unexpected operator: "${lexeme}".`);
    }
  }

  private parseUnaryOperator(lexeme: string, prefix: boolean): Operator {
    switch (lexeme) {
      case '++':
        return prefix ? Operator.UnaryIncrement : Operator.Increment;
      case '--':
        return prefix ? Operator.UnaryDecrement : Operator.Decrement;
      case '-':
        return Operator.UnaryNegative;
      case '+':
        return Operator.UnaryPositive;
      case '!':
        return Operator.UnaryNegation;
      default:
        // TODO: Test.
        /* istanbul ignore next */
        throw new SyntaxError(`Unexpected operator: "${lexeme}".`);
    }
  }

  /**
   * Returns whether any of the provided @param kinds are seen in order.
   */
  private match(...kinds: TokenKind[]): boolean {
    return kinds.some(e => {
      if (this.check(e)) {
        this.advance();
        return true;
      }
      return false;
    });
  }

  /**
   * Returns whether the next token is of type @param kind.
   */
  private check(kind: TokenKind): boolean {
    return this.hasNext ? this.peek().kind === kind : false;
  }

  /**
   * Returns the next token.
   */
  private advance(): IToken {
    if (this.hasNext) {
      this.position++;
    }
    return this.peek(-1);
  }

  /**
   * Returns whether at least one more token remains for parsing.
   */
  private get hasNext(): boolean {
    return this.tokens[this.position].kind !== TokenKind.EOF;
  }

  /**
   * Returns the token at @member position + @param offset.
   */
  private peek(offset = 0): IToken {
    return this.tokens[this.position + offset];
  }
}
