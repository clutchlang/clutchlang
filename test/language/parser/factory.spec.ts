// tslint:disable:no-magic-numbers

import { TokenKind } from '../../../src/language/lexer';
import { AstNodeFactory, Operator } from '../../../src/language/parser';

describe('AstNodeFactory', () => {
  const factory = new AstNodeFactory();

  describe('<Expression>', () => {
    const a = factory.createSimpleName({
      comments: [],
      kind: TokenKind.IDENTIFIER,
      lexeme: 'a',
      offset: 0,
    });
    const b = factory.createSimpleName({
      comments: [],
      kind: TokenKind.IDENTIFIER,
      lexeme: 'b',
      offset: 0,
    });

    it('should create BinaryExpression', () => {
      const $plus = {
        comments: [],
        kind: TokenKind.PLUS,
        lexeme: '+',
        offset: 0,
      };
      const expr = factory.createBinaryExpression(a, Operator.Add, $plus, b);
      expect(expr.firstToken).toBe(a.firstToken);
      expect(expr.lastToken).toBe(b.lastToken);
      expect(expr.left).toBe(a);
      expect(expr.operator).toBe(Operator.Add);
      expect(expr.right).toBe(b);
      expect(expr.target).toBe(a);
    });

    it('should create prefix UnaryExpression', () => {
      const $negate = {
        comments: [],
        kind: TokenKind.NEGATE,
        lexeme: '!',
        offset: 0,
      };
      const expr = factory.createUnaryExpression(
        a,
        Operator.UnaryNegative,
        $negate,
        true
      );
      expect(expr.firstToken).toBe($negate);
      expect(expr.lastToken).toBe(a.lastToken);
      expect(expr.operator).toEqual(Operator.UnaryNegation);
      expect(expr.target).toBe(a);
    });

    it('should create postfix UnaryExpression', () => {
      const $accessor = {
        comments: [],
        kind: TokenKind.PERIOD,
        lexeme: '!',
        offset: 0,
      };
      const expr = factory.createUnaryExpression(
        a,
        Operator.Accessor,
        $accessor,
        false
      );
      expect(expr.firstToken).toBe(a.firstToken);
      expect(expr.lastToken).toBe($accessor);
      expect(expr.operator).toEqual(Operator.Accessor);
      expect(expr.target).toBe(a);
    });

    describe('should create IfExpression', () => {
      const $if = {
        comments: [],
        kind: TokenKind.IF,
        lexeme: 'if',
        offset: 0,
      };

      it('', () => {
        const expr = factory.createIfExpression($if, a, b);
        expect(expr.body).toEqual(b);
        expect(expr.condition).toEqual(a);
        expect(expr.elseBody).toBeUndefined();
        expect(expr.elseToken).toBeUndefined();
        expect(expr.firstToken).toEqual($if);
        expect(expr.ifToken).toBe($if);
        expect(expr.lastToken).toBe(b.lastToken);
      });

      it('with else', () => {
        const $else = {
          comments: [],
          kind: TokenKind.ELSE,
          lexeme: 'else',
          offset: 0,
        };
        const expr = factory.createIfExpression($if, a, b, $else, a);
        expect(expr.body).toEqual(b);
        expect(expr.condition).toEqual(a);
        expect(expr.elseBody).toEqual(a);
        expect(expr.elseToken).toEqual($else);
        expect(expr.firstToken).toEqual($if);
        expect(expr.ifToken).toBe($if);
        expect(expr.lastToken).toBe(a.lastToken);
      });
    });
  });

  describe('LiteralBoolean', () => {
    it('should evaluate true', () => {
      const token = {
        comments: [],
        kind: TokenKind.TRUE,
        lexeme: 'true',
        offset: 0,
      };
      const $true = factory.createLiteralBoolean(token);
      expect($true.firstToken).toBe(token);
      expect($true.lastToken).toBe(token);
      expect($true.value).toBe(true);
    });

    it('should evaluate false', () => {
      const token = {
        comments: [],
        kind: TokenKind.TRUE,
        lexeme: 'false',
        offset: 0,
      };
      const $true = factory.createLiteralBoolean(token);
      expect($true.firstToken).toBe(token);
      expect($true.lastToken).toBe(token);
      expect($true.value).toBe(false);
    });
  });

  describe('LiteralNumber', () => {
    it('should evaluate int', () => {
      const token = {
        comments: [],
        kind: TokenKind.NUMBER,
        lexeme: '1',
        offset: 0,
      };
      const $1 = factory.createLiteralNumber(token);
      expect($1.firstToken).toBe(token);
      expect($1.lastToken).toBe(token);
      expect($1.value).toBe(1);
    });

    it('should evaluate float', () => {
      const token = {
        comments: [],
        kind: TokenKind.NUMBER,
        lexeme: '1.5',
        offset: 0,
      };
      const $1 = factory.createLiteralNumber(token);
      expect($1.firstToken).toBe(token);
      expect($1.lastToken).toBe(token);
      expect($1.value).toBe(1.5);
    });

    it('should evaluate hex', () => {
      const token = {
        comments: [],
        kind: TokenKind.NUMBER,
        lexeme: '0xFFF',
        offset: 0,
      };
      const $0xFFF = factory.createLiteralNumber(token);
      expect($0xFFF.firstToken).toBe(token);
      expect($0xFFF.lastToken).toBe(token);
      expect($0xFFF.value).toBe(0xfff);
    });

    it('should evaluate exponential', () => {
      const token = {
        comments: [],
        kind: TokenKind.NUMBER,
        lexeme: '2e6',
        offset: 0,
      };
      const $2e6 = factory.createLiteralNumber(token);
      expect($2e6.firstToken).toBe(token);
      expect($2e6.lastToken).toBe(token);
      expect($2e6.value).toBe(2e6);
    });
  });

  describe('LiteralString', () => {
    it('should evaluate an empty string', () => {
      const token = {
        comments: [],
        kind: TokenKind.STRING,
        lexeme: '',
        offset: 0,
      };
      const empty = factory.createLiteralString(token);
      expect(empty.firstToken).toBe(token);
      expect(empty.lastToken).toBe(token);
      expect(empty.value).toBe('');
    });

    it('should evaluate a single-line string', () => {
      const token = {
        comments: [],
        kind: TokenKind.STRING,
        lexeme: 'Hello',
        offset: 0,
      };
      const empty = factory.createLiteralString(token);
      expect(empty.firstToken).toBe(token);
      expect(empty.lastToken).toBe(token);
      expect(empty.value).toBe('Hello');
    });

    it('should evaluate a single-line string with escaped newlines', () => {
      const token = {
        comments: [],
        kind: TokenKind.STRING,
        lexeme: String.raw`Hello\nWorld`,
        offset: 0,
      };
      const empty = factory.createLiteralString(token);
      expect(empty.firstToken).toBe(token);
      expect(empty.lastToken).toBe(token);
      expect(empty.value).toBe('Hello\nWorld');
    });

    it('should evaluate a multi-line string', () => {
      const token = {
        comments: [],
        kind: TokenKind.STRING,
        lexeme: `
        <html>
          <body></body>
        </html>
      `,
        offset: 0,
      };
      const empty = factory.createLiteralString(token);
      expect(empty.firstToken).toBe(token);
      expect(empty.lastToken).toBe(token);
      expect(empty.value).toBe('<html>\n  <body></body>\n</html>\n');
    });
  });

  it('SimpleName should implement AstNode', () => {
    const token = {
      comments: [],
      kind: TokenKind.IDENTIFIER,
      lexeme: 'fooBar',
      offset: 0,
    };
    const fooBar = factory.createSimpleName(token);
    expect(fooBar.firstToken).toBe(token);
    expect(fooBar.lastToken).toBe(token);
    expect(fooBar.name).toBe('fooBar');
  });
});
