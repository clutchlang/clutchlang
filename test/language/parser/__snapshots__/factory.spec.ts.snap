// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstNodeFactory <Expression> should create BinaryExpression 1`] = `
"BinaryExpression:
  Left:
    <#a>
  Operator: +
  Right:
    <#b>
"
`;

exports[`AstNodeFactory <Expression> should create GroupExpression 1`] = `
"GroupExpression:
  Expression:
    <#a>
"
`;

exports[`AstNodeFactory <Expression> should create IfExpression  1`] = `
"IfExpression:
  If:
    <#a>
  Body:
    <#b>
"
`;

exports[`AstNodeFactory <Expression> should create IfExpression with else 1`] = `
"IfExpression:
  If:
    <#a>
  Body:
    <#b>
"
`;

exports[`AstNodeFactory <Expression> should create postfix UnaryExpression 1`] = `
"UnaryExpression:
  Target:
    <#a>
  Operator: ! (Postfix)
"
`;

exports[`AstNodeFactory <Expression> should create prefix UnaryExpression 1`] = `
"UnaryExpression:
  Target:
    <#a>
  Operator: ! (Prefix)
"
`;

exports[`AstNodeFactory LiteralBoolean should evaluate false 1`] = `
"<false>
"
`;

exports[`AstNodeFactory LiteralBoolean should evaluate true 1`] = `
"<true>
"
`;

exports[`AstNodeFactory LiteralNumber should evaluate exponential 1`] = `
"<2000000>
"
`;

exports[`AstNodeFactory LiteralNumber should evaluate float 1`] = `
"<1.5>
"
`;

exports[`AstNodeFactory LiteralNumber should evaluate hex 1`] = `
"<4095>
"
`;

exports[`AstNodeFactory LiteralNumber should evaluate int 1`] = `
"<1>
"
`;

exports[`AstNodeFactory LiteralString should evaluate an empty string 1`] = `
"<>
"
`;

exports[`AstNodeFactory SimpleName should implement AstNode 1`] = `
"<#fooBar>
"
`;

exports[`AstNodeFactory Statement should create a let statement 1`] = `
"VariableStatement:
  Assign:
    <#b>
"
`;

exports[`AstNodeFactory Statement should create a return statement 1`] = `
"JumpStatement:
  Return:
    <#a>
"
`;

exports[`AstNodeFactory should create a file root 1`] = `
"FileRoot:
  Elements:
    FunctionDeclaration:
      Name:
        <#a>
      Parameters:
      Body:
        <#b>
"
`;

exports[`AstNodeFactory should create a function 1`] = `
"FunctionDeclaration:
  Name:
    <#a>
  Parameters:
    <#b>
  Body:
    <#b>
"
`;

exports[`AstNodeFactory should create an invoke expression 1`] = `
"InvokeExpression:
  Target:
  Parameters:
"
`;
